[JSC] Enhance error message for for-of loops without Symbol.iterator

iidmsa opened the pull request:
#### 0051bbf2491fd2b54a5f2465a7e932e47ec184a0
<pre>
[JSC] Enhance error message for for-of loops without Symbol.iterator
<a href="https://bugs.webkit.org/show_bug.cgi?id=157041">https://bugs.webkit.org/show_bug.cgi?id=157041</a>
<a href="https://rdar.apple.com/problem/159814766">rdar://problem/159814766</a>

Reviewed by Keith Miller.

This adds a check to verify that Symbol.iterator property is callable
before attempting to use it. If it&apos;s not callable (undefined, null, or not a function),
we now throw a clearer TypeError message. This avoids adding error checks at the
bytecode generation stage and instead handles errors at runtime.

* JSTests/stress/iterator-error-messages-consistent.js:
(runTest):
* Source/JavaScriptCore/runtime/CommonSlowPaths.cpp:
(JSC::iteratorOpenTryFastImpl):
* Source/JavaScriptCore/runtime/IteratorOperations.cpp:
(JSC::validateIterable):
(JSC::throwIteratorError):
* Source/JavaScriptCore/runtime/IteratorOperations.h:
(JSC::forEachInIterable):
* JSTests/stress/iterator-error-messages-consistent.js: Added.
(shouldBe):
(shouldThrow):
(iterate):
(runTest):
(test):
(testFromEntries):
(testDestructuring):
(testSpread):
* JSTests/stress/object-from-entries.js:
(): Deleted.
* LayoutTests/inspector/unit-tests/array-utilities-expected.txt:
* LayoutTests/inspector/unit-tests/set-utilities-expected.txt:
* LayoutTests/js/let-syntax-expected.txt:
* Source/JavaScriptCore/runtime/CommonSlowPaths.cpp:
(JSC::iteratorOpenTryFastImpl):
* Source/JavaScriptCore/runtime/IteratorOperations.cpp:
(JSC::validateIterable):
(JSC::throwIteratorError):
* Source/JavaScriptCore/runtime/IteratorOperations.h:
(JSC::forEachInIterable):

Canonical link: <a href="https://commits.webkit.org/300154@main">https://commits.webkit.org/300154@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b3f90beea779b89b34ed104cf6afe74edcfb4fa2

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/121457 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/41154 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/31813 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/127909 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/73540 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/41856 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/49733 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/92251 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/61374 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/124409 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/33419 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/108805 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/72927 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/32429 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/26966 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/71478 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/113587 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/102908 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/27140 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/130732 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/119977 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/48385 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/36789 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/100840 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/48753 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/105025 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/100747 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25556 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/46158 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/24237 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/45081 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/48248 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/53956 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/150139 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/47715 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/38291 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/51061 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/49397 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation reviewed (changes requested):
Increasing bytecode size cause significant regression un perf and memory so, not OK.

Constellation commented on Source/JavaScriptCore/runtime/IteratorOperations.cpp:
```diff
@@ -247,8 +247,31 @@ IterationRecord iteratorDirect(JSGlobalObject* globalObject, JSValue object)
+IterationMode getIterationMode(VM& vm, JSGlobalObject* globalObject, JSValue iterable, JSValue symbolIterator)
 {
+    if (!symbolIterator.isCallable()) {
+        auto scope = DECLARE_THROW_SCOPE(vm);
```
This needs to be placed at the prologue of this function.

iidmsa replied:
Done.

Constellation commented on Source/JavaScriptCore/runtime/IteratorOperations.cpp:
```diff
@@ -247,8 +247,31 @@ IterationRecord iteratorDirect(JSGlobalObject* globalObject, JSValue object)
-IterationMode getIterationMode(VM&, JSGlobalObject* globalObject, JSValue iterable, JSValue symbolIterator)
+IterationMode getIterationMode(VM& vm, JSGlobalObject* globalObject, JSValue iterable, JSValue symbolIterator)
 {
+    if (!symbolIterator.isCallable()) {
```
Let's put `[[unlikely]]`.

iidmsa replied:
Done.

Constellation commented on Source/JavaScriptCore/runtime/IteratorOperations.cpp:
```diff
@@ -247,8 +247,31 @@ IterationRecord iteratorDirect(JSGlobalObject* globalObject, JSValue object)
 
-IterationMode getIterationMode(VM&, JSGlobalObject* globalObject, JSValue iterable, JSValue symbolIterator)
+IterationMode getIterationMode(VM& vm, JSGlobalObject* globalObject, JSValue iterable, JSValue symbolIterator)
 {
```
Now, this function can throw an error. So let's first put `JSGlobalObject*` to the first parameter.
Also, all the caller of this function needs to do exception check. So these code is necessary.

Constellation commented on Source/JavaScriptCore/runtime/CommonSlowPaths.cpp:
```diff
@@ -798,7 +798,7 @@ ALWAYS_INLINE UGPRPair iteratorOpenTryFastImpl(VM& vm, JSGlobalObject* globalObj
     auto& iterator = GET(bytecode.m_iterator);
 
-    if (getIterationMode(vm, globalObject, iterable, symbolIterator) == IterationMode::FastArray) {
+    if (getIterationMode(globalObject, vm, iterable, symbolIterator) == IterationMode::FastArray) {
```
This needs to do exception check.

iidmsa replied:
Done. Fixed.

Constellation commented on Source/JavaScriptCore/runtime/CommonSlowPaths.cpp:
```diff
@@ -791,19 +791,25 @@ JSC_DEFINE_COMMON_SLOW_PATH(slow_path_is_constructor)
 
-    if (getIterationMode(vm, globalObject, iterable, symbolIterator) == IterationMode::FastArray) {
+    auto iterationMode = getIterationMode(globalObject, vm, iterable, symbolIterator);
+    RETURN_IF_EXCEPTION(scope, encodeResult(pc, reinterpret_cast<void*>(static_cast<uintptr_t>(IterationMode::Generic))));
```
Using RETURN_IF_EXCEPTION is probably not correct since this code is for LLInt. Please read the code around this file and align to what they are doing.

iidmsa replied:
Resorted to check `throwScope.exception()`

Constellation commented on Source/JavaScriptCore/runtime/CommonSlowPaths.cpp:
```diff
@@ -791,19 +791,25 @@ JSC_DEFINE_COMMON_SLOW_PATH(slow_path_is_constructor)
         GET(bytecode.m_next) = JSValue();
         auto* iteratedObject = jsCast<JSObject*>(iterable);
         iterator = JSArrayIterator::create(vm, globalObject->arrayIteratorStructure(), iteratedObject, IterationKind::Values);
+        RETURN_IF_EXCEPTION(scope, encodeResult(pc, reinterpret_cast<void*>(static_cast<uintptr_t>(IterationMode::Generic))));
```
Ditto.

iidmsa replied:
Resorted to check `throwScope.exception()`

Constellation reviewed (commented):
Looks like EWS test are failing.

sosukesuzuki commented on Source/JavaScriptCore/runtime/IteratorOperations.cpp:
```diff
@@ -247,8 +247,32 @@ IterationRecord iteratorDirect(JSGlobalObject* globalObject, JSValue object)
+        else if (iterable.isNumber())
+            errorMessage = "number is not iterable"_s;
+        else
+            errorMessage = "value is not iterable"_s;
```
Is these error message consistent even after DFG or FTL? I think we should add tests to ensure error messages are consistent in DFG and FTL as well.

iidmsa replied:
Added a new test case that validates consistent behavior across both DFG and FTL execution tiers.

sosukesuzuki commented on Source/JavaScriptCore/runtime/IteratorOperations.cpp:
```diff
@@ -247,8 +247,32 @@ IterationRecord iteratorDirect(JSGlobalObject* globalObject, JSValue object)
+        else if (iterable.isBoolean())
+            errorMessage = iterable.asBoolean() ? "true is not iterable"_s : "false is not iterable"_s;
+        else if (iterable.isString())
+            errorMessage = "string is not iterable"_s;
```
String is an iterable?

iidmsa replied:
Reverted.

sosukesuzuki commented on Source/JavaScriptCore/runtime/IteratorOperations.cpp:
```diff
@@ -247,8 +247,32 @@ IterationRecord iteratorDirect(JSGlobalObject* globalObject, JSValue object)
+        else if (iterable.isUndefined())
+            errorMessage = "undefined is not iterable"_s;
+        else if (iterable.isObject())
+            errorMessage = "{} is not iterable"_s;
```
I don't think this message is not user friendly

iidmsa replied:
I was just trying to align with Chrome's V8 behavior.

sosukesuzuki reviewed (commented):
I think we need to create a new function that takes an iterable and validates it separately for runtime and LLInt, rather than throwing a `TypeError` in `getIterationMode`. Then call that function before `getIterationMode`.

sosukesuzuki commented on JSTests/stress/iterator-error-messages-consistent.js:
```diff
@@ -0,0 +1,81 @@
+}
+
+function runTest(value, expectedError) {
+    for (let i = 0; i < 10000; i++) {
```
You should use `testLoopCount`

iidmsa replied:
Done.

sosukesuzuki commented on JSTests/stress/iterator-error-messages-consistent.js:
```diff
@@ -0,0 +1,81 @@
+    shouldThrow(test, expectedError);
+}
+
+for (let i = 0; i < 10000; i++) {
```
ditto

iidmsa replied:
Done.

sosukesuzuki commented on JSTests/stress/iterator-error-messages-consistent.js:
```diff
@@ -0,0 +1,81 @@
+    shouldThrow(test, expectedError);
+}
+
+for (let i = 0; i < 10000; i++) {
```
ditto

iidmsa replied:
Done.

sosukesuzuki commented on JSTests/stress/iterator-error-messages-consistent.js:
```diff
@@ -0,0 +1,81 @@
+    shouldThrow(test, expectedError);
+}
+
+for (let i = 0; i < 10000; i++) {
```
ditto

iidmsa replied:
Done.

sosukesuzuki commented on Source/JavaScriptCore/runtime/IteratorOperations.cpp:
```diff
@@ -247,6 +247,27 @@ IterationRecord iteratorDirect(JSGlobalObject* globalObject, JSValue object)
     return { object, object.get(globalObject, globalObject->vm().propertyNames->next) };
 }
 
+void validateIterable(JSGlobalObject* globalObject, VM& vm, JSValue iterable, JSValue symbolIterator)
```
I don't think `validateIterable` should not throw an error. Should return a something like a enum or something like:

```cpp
enum class IterableValidationResult : uint8_t {
  Valid,
  InvalidString,
  /* ...other types */
}
```

And should throw TypeError in each runtime and LLInt side

What do you think

iidmsa replied:
Hm...Sounds interesting. New Design applied.

Constellation commented on Source/JavaScriptCore/runtime/CommonSlowPaths.cpp:
```diff
@@ -798,7 +798,15 @@ ALWAYS_INLINE UGPRPair iteratorOpenTryFastImpl(VM& vm, JSGlobalObject* globalObj
+    if (validationResult != IterableValidationResult::Valid) [[unlikely]] {
+        throwIteratorError(globalObject, vm, validationResult, iterable);
+        return encodeResult(pc, reinterpret_cast<void*>(static_cast<uintptr_t>(IterationMode::Generic)));
+    }
```
This looks not correct. Please read `THROW` etc. code carefully and align this to that.

iidmsa replied:
Done. Fixed.

Constellation commented on Source/JavaScriptCore/runtime/IteratorOperations.h:
```diff
@@ -170,7 +184,14 @@ void forEachInIterable(JSGlobalObject& globalObject, JSObject* iterable, JSValue
+    auto validationResult = validateIterable(vm, iterable, iteratorMethod);
+    if (validationResult != IterableValidationResult::Valid) {
+        createIteratorError(&globalObject, vm, validationResult, iterable);
+        return;
```
Returning an error object is not aligned to JSC's convention.
If `scope` is defined and if it is throwing an error, you need to throw an error.

iidmsa replied:
Fixed.

Constellation commented on Source/JavaScriptCore/runtime/CommonSlowPaths.cpp:
```diff
@@ -815,23 +821,20 @@ ALWAYS_INLINE UGPRPair iteratorOpenTryFastImpl(VM& vm, JSGlobalObject* globalObj
 
 JSC_DEFINE_COMMON_SLOW_PATH(iterator_open_try_fast_narrow)
 {
-    // Don't set PC; we can't throw and it's relatively slow.
```
So, this comment is saying we cannot throw here. Why? Can you investigate and ensure that it is fine to throw? Or maybe it is not allowed to throw an error.

kmiller68 replied:
Yeah, this seems wrong. We don't check for an exception in the interpreter/JIT when we return.

kmiller68 commented on Source/JavaScriptCore/runtime/CommonSlowPaths.cpp:
```diff
@@ -798,7 +798,13 @@ ALWAYS_INLINE UGPRPair iteratorOpenTryFastImpl(VM& vm, JSGlobalObject* globalObj
-    if (getIterationMode(vm, globalObject, iterable, symbolIterator) == IterationMode::FastArray) {
+    auto validationResult = validateIterable(vm, iterable, symbolIterator);
+    if (validationResult != IterableValidationResult::Valid) [[unlikely]]
+        THROW(createIteratorError(globalObject, vm, validationResult, iterable));
```
We should probably only check this after getting `getIterationMode` since we'll only have an error if that returned Generic anyway. This avoids these checks for the fastest paths.

kmiller68 commented on Source/JavaScriptCore/runtime/IteratorOperations.h:
```diff
@@ -170,7 +184,14 @@ void forEachInIterable(JSGlobalObject& globalObject, JSObject* iterable, JSValue
+    }
+
+    auto IterationMode = getIterationMode(vm, &globalObject, iterable, iteratorMethod);
+    if (IterationMode == IterationMode::FastArray) {
```
Ditto on checking only on Generic.

Constellation commented on Source/JavaScriptCore/runtime/IteratorOperations.cpp:
```diff
@@ -247,6 +247,61 @@ IterationRecord iteratorDirect(JSGlobalObject* globalObject, JSValue object)
+    return IterableValidationResult::Valid;
+}
+
+JSObject* createIteratorError(JSGlobalObject* globalObject, VM&, IterableValidationResult result, JSValue iterable)
```
Why not just returning `ASCIILiteral` as a message instead of creating an error object here.

Constellation commented on Source/JavaScriptCore/runtime/IteratorOperations.cpp:
```diff
@@ -247,6 +247,61 @@ IterationRecord iteratorDirect(JSGlobalObject* globalObject, JSValue object)
+        break;
+    }
+
+    return createTypeError(globalObject, errorMessage);
```
Why not just returning a string message and using `throwTypeError` with the message in the caller side?

Constellation commented on Source/JavaScriptCore/runtime/IteratorOperations.h:
```diff
@@ -59,6 +59,20 @@ JS_EXPORT_PRIVATE IterationRecord iteratorDirect(JSGlobalObject*, JSValue);
+};
+
+JS_EXPORT_PRIVATE IterableValidationResult validateIterable(VM&, JSValue iterable, JSValue symbolIterator);
+JS_EXPORT_PRIVATE JSObject* createIteratorError(JSGlobalObject*, VM&, IterableValidationResult, JSValue iterable);
```
We should just get ASCIILiteral instead of creating an error object.

Constellation commented on Source/JavaScriptCore/runtime/IteratorOperations.h:
```diff
@@ -187,6 +202,12 @@ void forEachInIterable(JSGlobalObject& globalObject, JSObject* iterable, JSValue
 
+    auto validationResult = validateIterable(vm, iterable, iteratorMethod);
+    if (validationResult != IterableValidationResult::Valid) [[unlikely]] {
+        throwException(&globalObject, scope, createIteratorError(&globalObject, vm, validationResult, iterable));
```
We should use `throwTypeError`.

Constellation commented on Source/JavaScriptCore/runtime/CommonSlowPaths.cpp:
```diff
@@ -808,30 +809,33 @@ ALWAYS_INLINE UGPRPair iteratorOpenTryFastImpl(VM& vm, JSGlobalObject* globalObj
 
+    auto validationResult = validateIterable(vm, iterable, symbolIterator);
+    if (validationResult != IterableValidationResult::Valid) [[unlikely]] {
+        throwScope.throwException(globalObject, createIteratorError(globalObject, vm, validationResult, iterable));
```
You should use `throwTypeError`.

Constellation commented on Source/JavaScriptCore/llint/LowLevelInterpreter64.asm:
```diff
@@ -3304,12 +3304,12 @@ llintOpWithMetadata(op_iterator_open, OpIteratorOpen, macro (size, get, dispatch
     dispatch()
 
 .iteratorOpenGeneric:
+    btpz r0, .iteratorOpenException
```
You should check this before jumping to here. (Ideally, just after the call). There are many cases exception can be thrown.

kmiller68 replied:
Oh I requested this behavior since we should just return Generic from the operation when an exception is thrown. It saves a branch for fast path iterators.

Constellation commented on Source/JavaScriptCore/llint/LowLevelInterpreter32_64.asm:
```diff
@@ -3117,12 +3117,12 @@ llintOpWithMetadata(op_iterator_open, OpIteratorOpen, macro (size, get, dispatch
     dispatch()
 
 .iteratorOpenGeneric:
+    btpz r0, .iteratorOpenException
```
Ditto.

kmiller68 reviewed (approved):
r=me.

iidmsa committed:
[JSC] Enhance error message for for-of loops without Symbol.iterator
https://bugs.webkit.org/show_bug.cgi?id=157041
rdar://problem/159814766

Reviewed by Keith Miller.

This adds a check to verify that Symbol.iterator property is callable
before attempting to use it. If it's not callable (undefined, null, or not a function),
we now throw a clearer TypeError message. This avoids adding error checks at the
bytecode generation stage and instead handles errors at runtime.

* JSTests/stress/iterator-error-messages-consistent.js:
(runTest):
* Source/JavaScriptCore/runtime/CommonSlowPaths.cpp:
(JSC::iteratorOpenTryFastImpl):
* Source/JavaScriptCore/runtime/IteratorOperations.cpp:
(JSC::validateIterable):
(JSC::throwIteratorError):
* Source/JavaScriptCore/runtime/IteratorOperations.h:
(JSC::forEachInIterable):
* JSTests/stress/iterator-error-messages-consistent.js: Added.
(shouldBe):
(shouldThrow):
(iterate):
(runTest):
(test):
(testFromEntries):
(testDestructuring):
(testSpread):
* JSTests/stress/object-from-entries.js:
(): Deleted.
* LayoutTests/inspector/unit-tests/array-utilities-expected.txt:
* LayoutTests/inspector/unit-tests/set-utilities-expected.txt:
* LayoutTests/js/let-syntax-expected.txt:
* Source/JavaScriptCore/runtime/CommonSlowPaths.cpp:
(JSC::iteratorOpenTryFastImpl):
* Source/JavaScriptCore/runtime/IteratorOperations.cpp:
(JSC::validateIterable):
(JSC::throwIteratorError):
* Source/JavaScriptCore/runtime/IteratorOperations.h:
(JSC::forEachInIterable):

Canonical link: https://commits.webkit.org/300154@main
